libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -642,49 +645,53 @@ BeaconSetupService::State BeaconSetupService::OnWaitForReconstructionShares()
-  {
-    // Process reconstruction shares. Reconstruction shares from non-qual members
-    // or people in qual complaints should not be considered
-    for (auto const &share : reconstruction_shares_received_)
+    // Reconstruction can take some time - best to do it ASAP
     {
-      MuddleAddress from = share.first;
-      if (qual_complaints_manager_.FindComplaint(from) ||
-          beacon_->manager.qual().find(from) == beacon_->manager.qual().end())
+      // Process reconstruction shares. Reconstruction shares from non-qual members
+      // or people in qual complaints should not be considered
+      for (auto const &share : reconstruction_shares_received_)
       {
-        FETCH_LOG_WARN(LOGGING_NAME, "Node ", beacon_->manager.cabinet_index(),
-                       " received message from invalid sender. Discarding.");
-        continue;
-      }
-      for (auto const &elem : share.second)
-      {
-        // If address of node who's shares is being exposed is not in qual then
-        // do not process
-        if (qual.find(elem.first) != qual.end())
+        MuddleAddress from = share.first;
+        if (qual_complaints_manager_.FindComplaint(from) ||
+            beacon_->manager.qual().find(from) == beacon_->manager.qual().end())
+        {
+          FETCH_LOG_WARN(LOGGING_NAME, "Node ", beacon_->manager.cabinet_index(),
+                         " received message from invalid sender. Discarding.");
+          continue;
+        }
+        for (auto const &elem : share.second)
         {
           beacon_->manager.VerifyReconstructionShare(from, elem);
         }
       }
-    }
 
-    // Reset if reconstruction fails as this breaks the initial assumption on the
-    // number of Byzantine nodes
-    if (!beacon_->manager.RunReconstruction())
-    {
-      FETCH_LOG_WARN(LOGGING_NAME, "Node: ", beacon_->manager.cabinet_index(),
-                     " DKG failed due to reconstruction failure. Resetting.");
-      SetTimeToProceed(State::RESET);
-      return State::RESET;
-    }
-    else
-    {
-      beacon_->manager.ComputePublicKeys();
+      // Reset if reconstruction fails as this breaks the initial assumption on the
+      // number of Byzantine nodes
+      if (!beacon_->manager.RunReconstruction())
+      {
+        FETCH_LOG_WARN(LOGGING_NAME, "Node: ", beacon_->manager.cabinet_index(),
+                       " DKG failed due to reconstruction failure. Resetting.");
+        SetTimeToProceed(State::RESET);
+        beacon_dkg_state_failed_on_->set(static_cast<uint64_t>(state_machine_->state()));
+        return State::RESET;
+      }
+      else
+      {
+        beacon_->manager.ComputePublicKeys();
+      }
+
+      SetTimeToProceed(State::DRY_RUN_SIGNING);
+      return State::DRY_RUN_SIGNING;
     }
+  }
 
-    SetTimeToProceed(State::DRY_RUN_SIGNING);
-    return State::DRY_RUN_SIGNING;
+  if (timer_to_proceed_.HasExpired())
+  {
+    FETCH_LOG_INFO(LOGGING_NAME, "Timed out during reconstruction.");
```

Comment:
I don't think this if statement is correct. Currently, if a node is missing one reconstruction share it resets, when it could in fact proceed onto running reconstruction. As I mentioned the if statement should be if the timer has expired or one has collected cabinet_size - 1 reconstruction shares then do the verification and run reconstruction etc. If compute public keys is taking a long time then it should be moved into its own state.

Reply:
fixed